Sorting language pragmas should be more intelligent

This was mentioned in #216, but it is a separate issue.

>The problem is that a list of language extensions is not always a list of commutative elements, however it often is a list of commutative elements and users will likely expect the alphabetized behavior in the latter case.

> If one were to take care to note which language extensions were implied by which other ones, we can determine which elements in the list of language extensions could not be commuted "past" one another when alphabetizing. If this bookkeepping was performed, we could indempotently output a mostly alphabetized list of language extensions when one or more elements cannot commute "past" each other or a completely alphabetized list when all elements can be commuted, preserving the module's semantics in either case.

> While this approach would require much more work to implement and maintain, it is the most correct option for an opinionated code styler.

IOW, we should take into account that some pragmas that disable extensions enabled as side-effect of enabling other extensions should be put after those extensions because sometimes the order of extension pragmas matters.

The linked table has only two `turnOff` entries:

```
(LangExt.RebindableSyntax, turnOff, LangExt.ImplicitPrelude)
(LangExt.StandaloneKindSignatures, turnOff, LangExt.CUSKs)
```

I think we can ignore `LANGUAGE CUSKs`, it seems to be very, very rare (only one example on GitHub). And funnily, you can't actually have `RebindableSyntax` and `ImplicitPrelude` together, no matter the order:

```haskell
{-# LANGUAGE RebindableSyntax #-}
{-# LANGUAGE ImplicitPrelude #-}

main = if True then print "hi" else print "bye"
```

```
$ ghc x.hs
[1 of 1] Compiling Main             ( x.hs, x.o )

x.hs:4:8: error: Not in scope: ‘ifThenElse’
  |
4 | main = if True then print "hi" else print "bye"
  |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

*This leaves only the case where the user says `{-# LANGUAGE NoX #-}` to disable some `X` enabled by a different extension earlier. This can be handled by putting all `NoX` pragmas into a separate section (my preferred option), or simply sorting them at the end of the list.*

Yes, we can no longer say "if it compiled before reformatting, it will compile after reformatting", but on the other hand we can prevent people from having to debug "why doesn't it compile even though I said NoX?" errors. Plus, I find myself somewhat persuaded by @mboes's comment:

> [...] the semantics of the source file, resulting from sorting the language pragmas, could well be the semantics the programmer intends, even if it's sometimes different from the input.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -38,6 +38,9 @@
 * Implemented correct handling of stack headers. [Issue
   393](https://github.com/tweag/ormolu/issues/393).
 
+* Sorting language pragmas cannot not change meaning of the input program
+  anymore. [Issue 404](https://github.com/tweag/ormolu/issues/404).
+
 ## Ormolu 0.0.1.0
 
 * Initial release.

--- a/data/examples/declaration/class/type-operators3-out.hs
+++ b/data/examples/declaration/class/type-operators3-out.hs
@@ -1,7 +1,7 @@
-{-# LANGUAGE NoStarIsType #-}
 {-# LANGUAGE PolyKinds #-}
 {-# LANGUAGE TypeFamilies #-}
 {-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE NoStarIsType #-}
 
 class PNum x where
   type (a :: x) * (b :: x)

--- a/data/examples/other/pragma-sorting-out.hs
+++ b/data/examples/other/pragma-sorting-out.hs
@@ -1,0 +1,12 @@
+{-# LANGUAGE NondecreasingIndentation #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE NoMonoLocalBinds #-}
+
+-- This gap is necessary for stylish Haskell not to re-arrange
+-- NoMonoLocalBinds before TypeFamilies
+
+module Foo
+  ( bar,
+  )
+where

--- a/data/examples/other/pragma-sorting.hs
+++ b/data/examples/other/pragma-sorting.hs
@@ -1,0 +1,10 @@
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeFamilies        #-}
+{-# LANGUAGE NondecreasingIndentation #-}
+-- This gap is necessary for stylish Haskell not to re-arrange
+-- NoMonoLocalBinds before TypeFamilies
+{-# LANGUAGE NoMonoLocalBinds    #-}
+
+module Foo (
+   bar
+  ) where

--- a/src/Ormolu/Printer/Meat/Pragma.hs
+++ b/src/Ormolu/Printer/Meat/Pragma.hs
@@ -8,18 +8,46 @@ module Ormolu.Printer.Meat.Pragma
   )
 where
 
+import Data.Char (isUpper)
+import Data.Maybe (listToMaybe)
 import qualified Data.Set as S
 import qualified Data.Text as T
 import Ormolu.Parser.Pragma (Pragma (..))
 import Ormolu.Printer.Combinators
 
-data PragmaTy = Language | OptionsGHC | OptionsHaddock
+-- | Pragma classification.
+data PragmaTy
+  = Language LanguagePragmaClass
+  | OptionsGHC
+  | OptionsHaddock
+  deriving (Eq, Ord)
+
+-- | Language pragma classification.
+--
+-- The order in which language pragmas are put in the input sometimes
+-- matters. This is because some language extensions can enable other
+-- extensions, yet the extensions coming later in the list have the ability
+-- to change it. So here we classify all extensions by assigning one of the
+-- four groups to them. Then we only sort inside of the groups.
+--
+-- 'Ord' instance of this data type is what affects the sorting.
+--
+-- See also: <https://github.com/tweag/ormolu/issues/404>
+data LanguagePragmaClass
+  = -- | All other extensions
+    Normal
+  | -- | Extensions starting with "No"
+    Disabling
+  | -- | Extensions that should go after everything else
+    Final
   deriving (Eq, Ord)
 
 p_pragmas :: [Pragma] -> R ()
 p_pragmas ps =
   let prepare = concatMap $ \case
-        PragmaLanguage xs -> (Language,) <$> xs
+        PragmaLanguage xs ->
+          let f x = (Language (classifyLanguagePragma x), x)
+           in f <$> xs
         PragmaOptionsGHC x -> [(OptionsGHC, x)]
         PragmaOptionsHaddock x -> [(OptionsHaddock, x)]
    in mapM_ (uncurry p_pragma) (S.toAscList . S.fromList . prepare $ ps)
@@ -28,10 +56,26 @@ p_pragma :: PragmaTy -> String -> R ()
 p_pragma ty c = do
   txt "{-# "
   txt $ case ty of
-    Language -> "LANGUAGE"
+    Language _ -> "LANGUAGE"
     OptionsGHC -> "OPTIONS_GHC"
     OptionsHaddock -> "OPTIONS_HADDOCK"
   space
   txt (T.pack c)
   txt " #-}"
   newline
+
+-- | Classify a 'LanguagePragma'.
+classifyLanguagePragma :: String -> LanguagePragmaClass
+classifyLanguagePragma = \case
+  "ImplicitPrelude" -> Final
+  "CUSKs" -> Final
+  str ->
+    case splitAt 2 str of
+      ("No", rest) ->
+        case listToMaybe rest of
+          Nothing -> Normal
+          Just x ->
+            if isUpper x
+              then Disabling
+              else Normal
+      _ -> Normal

